### PR TITLE
Refreshing sample data

### LINKED ISSUE #384: Refresh sample data so that it is not based on real data
The [sample data](https://github.com/GOCDB/gocdb/tree/dev/lib/Doctrine/deploy/sampleData) looks like it was based on some real data at one point. It would be good to refresh this data so that this is no longer the case.

It would also be an oppurtunity to tweak the structure of the sample data, so that it better mirrors present data - i.e. contains multiple NGIs (one of which isn't under a project) like below (the sites should also have dummy services.):

```
        P1
      /    \    
     N1    N2      N3
    /  \   |       |
   S1  S2  S3      S4
```

Users should be given unique, but still dummy, emails so they are easier to differentiate (i.e. in dry runs of scripts that generate emails).

There exists some ["simple sample data"](https://github.com/GOCDB/gocdb/tree/dev/lib/Doctrine/deploy/simpleSampleData) which would be a good starting point to expand and replace the existing sample data.

As a follow on, the example API output used in the [docs](https://github.com/GOCDB/GOCDB.github.io/tree/main/api/read) should be updated to reflect the new sample data, see: https://github.com/GOCDB/GOCDB.github.io/issues/12.

### PR DESCRIPTION
**Pull request by rowan04**
Resolves #384
- Sanitised data so it's not based on real data
- Added multiple NGIs with sites and services
- Added more users with unique dummy emails

### CODE REVIEW

File: lib/Doctrine/deploy/AddNGIs.php [hunk 1]
```diff
@@ -51,6 +56,10 @@
             $securityEmail = (string) $value;
         }
 
+        if((string) $key == "OBJECT_ID") {
```
Comment (rowan04) [inline lib/Doctrine/deploy/AddNGIs.php:59]: This is the line causing the Codacy fail - however it is in the same format as the other if statements above. Should I fix?

Comment (gregcorbett) [reply lib/Doctrine/deploy/AddNGIs.php:59]: As it's a new line of code, could you address the issue. A seperate PR that fixes the issue for the file as a whole would also be most welcome :)

File: lib/Doctrine/deploy/sampleData/ServiceEndpoints.xml [hunk 2]
```diff
@@ -53,15 +53,15 @@
...
+	<SITENAME>Torch</SITENAME>
         <COUNTRY_NAME>Switzerland</COUNTRY_NAME>
         <COUNTRY_CODE>CH</COUNTRY_CODE>
         <ROC_NAME>NGI_CH</ROC_NAME>
         <URL></URL>
     </SERVICE_ENDPOINT>
     <SERVICE_ENDPOINT PRIMARY_KEY="43873G0">
         <PRIMARY_KEY>43873G0</PRIMARY_KEY>
-        <HOSTNAME>bdii.lhep.unibe.ch</HOSTNAME>
+        <HOSTNAME>bdii.lhep.unilu.grid.ch</HOSTNAME>
```
Comment (gregcorbett) [inline lib/Doctrine/deploy/sampleData/ServiceEndpoints.xml:64]: Hmmmm `unilu.ch` is a thing, maybe we could change `unilu.grid.ch` to something else - perhaps `torch.grid.ch`?

Comment (gregcorbett) [reply lib/Doctrine/deploy/sampleData/ServiceEndpoints.xml:64]: or perhaps something different so we can change the `unilu` sitename to something different to `torch`.

File: lib/Doctrine/deploy/sampleData/ServiceEndpoints.xml [hunk 3]
```diff
@@ -285,11 +285,385 @@
...
+        <NODE_MONITORED>Y</NODE_MONITORED>
+        <SITENAME>Izola MRF</SITENAME>
+        <COUNTRY_NAME>Slovenia</COUNTRY_NAME>
+        <COUNTRY_CODE>SI</COUNTRY_CODE>
+        <ROC_NAME>NGI_SI</ROC_NAME>
+        <URL></URL>
+    </SERVICE_ENDPOINT>
+    <SERVICE_ENDPOINT PRIMARY_KEY="10007G0">
+        <PRIMARY_KEY>10007G0</PRIMARY_KEY>
+        <HOSTNAME>uni-bdii.uni-lj.si</HOSTNAME>
```
Comment (gregcorbett) [inline lib/Doctrine/deploy/sampleData/ServiceEndpoints.xml:381]: looks like `uni-lj.si` is a thing, could this be changed?

Comment (gregcorbett) [reply lib/Doctrine/deploy/sampleData/ServiceEndpoints.xml:381]: and changed in other places as well.

File: lib/Doctrine/deploy/sampleData/ServiceEndpoints.xml [hunk 4]
```diff
@@ -285,11 +285,385 @@
...
+        <NODE_MONITORED>Y</NODE_MONITORED>
+        <SITENAME>Budapest Institute</SITENAME>
+        <COUNTRY_NAME>Hungary</COUNTRY_NAME>
+        <COUNTRY_CODE>HU</COUNTRY_CODE>
+        <ROC_NAME>NGI_HU</ROC_NAME>
+        <URL></URL>
+    </SERVICE_ENDPOINT>
+    <SERVICE_ENDPOINT PRIMARY_KEY="3133G0">
+        <PRIMARY_KEY>3133G0</PRIMARY_KEY>
+        <HOSTNAME>SRM.budapestinstitute-gridoperations.hu</HOSTNAME>
```
Comment (gregcorbett) [inline lib/Doctrine/deploy/sampleData/ServiceEndpoints.xml:483]: perhaps change 'institute' to some synonym - here and elsewhere.

File: lib/Doctrine/deploy/sampleData/ServiceEndpoints.xml [hunk 5]
```diff
@@ -285,11 +285,385 @@
...
+        <NODE_MONITORED>Y</NODE_MONITORED>
+        <SITENAME>Barcs-Drava STF</SITENAME>
+        <COUNTRY_NAME>Hungary</COUNTRY_NAME>
+        <COUNTRY_CODE>HU</COUNTRY_CODE>
+        <ROC_NAME>NGI_HU</ROC_NAME>
+        <URL></URL>
+    </SERVICE_ENDPOINT>
+    <SERVICE_ENDPOINT PRIMARY_KEY="9293G0">
+        <PRIMARY_KEY>9293G0</PRIMARY_KEY>
+        <HOSTNAME>site-bdii.danubeinstitute.hu</HOSTNAME>
```
Comment (gregcorbett) [inline lib/Doctrine/deploy/sampleData/ServiceEndpoints.xml:619]: perhaps change 'institute' to some synonym - here and elsewhere.

File: lib/Doctrine/deploy/sampleData/ServiceEndpoints.xml (hunk 2 above)
Comment (rowan04) [reply lib/Doctrine/deploy/sampleData/ServiceEndpoints.xml:64]: ok, I'll change the sitename and hence its' services

File: lib/Doctrine/deploy/sampleData/ServiceEndpoints.xml (hunk 3 above)
Comment (rowan04) [reply lib/Doctrine/deploy/sampleData/ServiceEndpoints.xml:381]: yep will change this sitename and related services too

File: lib/Doctrine/deploy/sampleData/ServiceEndpoints.xml (hunk 4 above)
Comment (rowan04) [reply lib/Doctrine/deploy/sampleData/ServiceEndpoints.xml:483]: ok

File: lib/Doctrine/deploy/sampleData/ServiceEndpoints.xml (hunk 5 above)
Comment (rowan04) [reply lib/Doctrine/deploy/sampleData/ServiceEndpoints.xml:619]: ok

File: lib/Doctrine/deploy/sampleData/ServiceEndpoints.xml [hunk 6]
```diff
@@ -285,11 +285,385 @@
...
+        <NODE_MONITORED>Y</NODE_MONITORED>
+        <SITENAME>Balaton RF</SITENAME>
+        <COUNTRY_NAME>Hungary</COUNTRY_NAME>
+        <COUNTRY_CODE>HU</COUNTRY_CODE>
+        <ROC_NAME>NGI_HU</ROC_NAME>
+        <URL></URL>
+    </SERVICE_ENDPOINT>
+    <SERVICE_ENDPOINT PRIMARY_KEY="7651G0">
+        <PRIMARY_KEY>7651G0</PRIMARY_KEY>
+        <HOSTNAME>srm.lakebalatonresearch.hu</HOSTNAME>
```
[the remark this answers is not in the document]
Comment (rowan04) [reply lib/Doctrine/deploy/sampleData/ServiceEndpoints.xml:517]: ok

**Comment by gregcorbett** [comment]
Looks good 😄 , the only other comment I have is - can you check the roles the Users have allign with the type of entity the role is over (see the "Role Action Map" page on GOCDB instances)?

i.e. Donna Butler has the "COD Staff" role over NGI_CH, but "COD Staff" is a Project entity role.

File: lib/Doctrine/deploy/sampleData/ServiceEndpoints.xml (hunk 6 above)
[the remark this answers is not in the document]
Comment (rowan04) [reply lib/Doctrine/deploy/sampleData/ServiceEndpoints.xml:517]: balaton is the name of the lake I made up this research facility to be next to, so i don't quite know what synonym i can give it? i was thinking of just editing the name from lakebalatonresearch.hu to something like lakebalatonpprf.hu or something, as maybe the research part is too generic. Or I could ditch balaton and call it something like lakesidePPRF. PPRF standing for particle physics research facility.

[the remark this answers is not in the document]
Comment (gregcorbett) [reply lib/Doctrine/deploy/sampleData/ServiceEndpoints.xml:517]: lakesidePPRF sounds good :)

[the remark this answers is not in the document]
Comment (rowan04) [reply lib/Doctrine/deploy/sampleData/ServiceEndpoints.xml:517]: got it!

**Comment by rowan04** [comment]
The commit just done is to do with the site names - I will hopefully be able to fix the user roles tomorrow morning

**Comment by gregcorbett** [comment]
> can you check the roles the Users have allign with the type of entity the role is over (see the "Role Action Map" page on GOCDB instances)?

I also now realise the sample User / Roles never made much sense (for example, it has always had Site level roles over NGIs), I assumed I had fixed this, but it seems that change has become trapped in #282 

It would be good to have a saner sample User / Role mapping given you are refreshing the sample data already.

The "Role Action Map" can be a little confusing (it's just confused me for a while resulting in me going down a platypus hole in the code).

Taking this as an example:

https://github.com/rowan04/gocdb/blob/8f5eb4772541a090ff04847d8022c666d21f1c16/lib/Doctrine/deploy/sampleData/UsersAndRoles.xml#L24-L28

"Site Administrator" is a Site level RoleType and should only be applied to a Site entity (i.e. the `ON_ENTITY` element should container a Site name, not an NGI name). 

Having a Site level role does grant actions over NGIs and Projects (which is how we end up with lines like this in the "Role Action Map".

image

but the "Site Administrator" only be applied to a site entity.

Does that make sense?

**Comment by rowan04** [comment]
> 
When I tried to make the ON_ENTITY element a sitename, it would produce an error saying something like "no NGI's found with _sitename_". That's what i tried to do originally for roles like site administrator but had to change them all back to NGI's.

**Comment by rowan04** [comment]
> > 
> 
> When I tried to make the ON_ENTITY element a sitename, it would produce an error saying something like "no NGI's found with _sitename_". That's what i tried to do originally for roles like site administrator but had to change them all back to NGI's.

Would you like me to reproduce the error?

**Comment by gregcorbett** [comment]
> > > 
> > 
> > 
> > When I tried to make the ON_ENTITY element a sitename, it would produce an error saying something like "no NGI's found with _sitename_". That's what i tried to do originally for roles like site administrator but had to change them all back to NGI's.
> 
> Would you like me to reproduce the error?

That's weird, yes please 😄

**Comment by rowan04** [comment]
> > > > 
> > > 
> > > 
> > > When I tried to make the ON_ENTITY element a sitename, it would produce an error saying something like "no NGI's found with _sitename_". That's what i tried to do originally for roles like site administrator but had to change them all back to NGI's.
> > 
> > 
> > Would you like me to reproduce the error?
> 
> That's weird, yes please 😄

nongis

**Comment by rowan04** [comment]
And I'm heading home now. I guess we'll return to this next year :)

**Comment by gregcorbett** [comment]
> I think it would be good to tackle #373 at the same time as this I think. Though we should probably just delete ServiceGroupRoles.xml and give additional servicegroup roles to the users in UsersAndRoles.xml.

Having just noticed #393 exists 🤭, lets not conflate the two - so let's not worry about servicegroups at all here :)

**Comment by rowan04** [comment]
> > > > > > > 
> > > > > > 
> > > > > > 
> > > > > > When I tried to make the ON_ENTITY element a sitename, it would produce an error saying something like "no NGI's found with _sitename_". That's what i tried to do originally for roles like site administrator but had to change them all back to NGI's.
> > > > > 
> > > > > 
> > > > > Would you like me to reproduce the error?
> > > > 
> > > > 
> > > > That's weird, yes please 😄
> > > 
> > > 
> > > nongis
> > 
> > 
> > this was the code torch
> 
> ah, looks like `<ENTITY_TYPE>group</ENTITY_TYPE>` is the culprit. Going a bit deeper it looks like UsersAndRoles.xml is based on an older version of the API output from GOCDB's `get_user` method 😢
> 
> Looking at the output now, `<ENTITY_TYPE>` is either `site`, `ngi`, `project`, `servicegroup`.
> 
> * AddSiteRoles.php looks like it already looks for `<ENTITY_TYPE>` equalling `site`: https://github.com/GOCDB/gocdb/blob/dev/lib/Doctrine/deploy/AddSiteRoles.php#L22
> * AddGroupRoles.php should probably be renamed AddNGIRoles.php and only focus on `<ENTITY_TYPE>` equalling `ngi`: https://github.com/GOCDB/gocdb/blob/dev/lib/Doctrine/deploy/AddGroupRoles.php#L17
> * AddEgiRoles.php should probably be generalised into AddProjectRoles.php and focus on and only focus on `<ENTITY_TYPE>` equalling `project` (and not assume it's working on the EGI project) see https://github.com/GOCDB/gocdb/blob/dev/lib/Doctrine/deploy/AddEgiRoles.php#L52 and https://github.com/GOCDB/gocdb/blob/dev/lib/Doctrine/deploy/AddEgiRoles.php#L57
> 
> I'm not sure what's going on with `<ENTITY_TYPE>` equalling `servicegroup`. We don't seem to have a AddServiceGroupRoles.php equivalent, but we do have some sample ServiceGroupRoles.. I think it would be good to tackle #373 at the same time as this I think. Though we should probably just delete ServiceGroupRoles.xml and give additional servicegroup roles to the users in UsersAndRoles.xml.

I see (I think!)

**Comment by rowan04** [comment]
> > I think it would be good to tackle #373 at the same time as this I think. Though we should probably just delete ServiceGroupRoles.xml and give additional servicegroup roles to the users in UsersAndRoles.xml.
> 
> Having just noticed #393 exists 🤭, lets not conflate the two - so let's not worry about servicegroups at all here :)

Ok.

File: lib/Doctrine/deploy/AddProjectRoles.php [hunk 7]
```diff
@@ -91,9 +88,13 @@
...
         }
 
-        $doctrineRole = new Role($roleType, $doctrineUser, $egi, 'STATUS_GRANTED');
-        $entityManager->persist($doctrineRole);
+        if ((string) $role->ON_ENTITY == "EGI") {
+            $doctrineRole = new Role($roleType, $doctrineUser, $egi, 'STATUS_GRANTED');
+            $entityManager->persist($doctrineRole);
+        } else {
+            continue;
+        }
```
Comment (gregcorbett) [inline lib/Doctrine/deploy/AddProjectRoles.php:96]: Can this be a look up on the underlying project entity and, regardless of whether it's EGI or another project, add the appropriate role?

Similar to https://github.com/rowan04/gocdb/blob/384-refresh_sample_data/lib/Doctrine/deploy/AddNGIRoles.php#L65-L78

Comment (gregcorbett) [reply lib/Doctrine/deploy/AddProjectRoles.php:96]: and ammend the docstring at the top of the file accordingly.

File: lib/Doctrine/deploy/AddProjectRoles.php [hunk 8]
```diff
@@ -5,33 +5,35 @@
 
 /**
  * AddEgiRoles.php: Loads a list of roles from the get_user PI
- * query output (XML), finds the EGI roles and inserts them into
- * the doctrine prototype.
+ * query output (XML), finds whether they are EGI roles or not
+ * and if they are, adds them to the EGI project and 
+ * insert them into the doctrine prototype.
+ * If they are not an EGI role, they are not added to a project
  * XML format is the output from get_roc_list PI query, e.g:
```
[the remark this answers is not in the document]
Comment (rowan04) [reply lib/Doctrine/deploy/AddProjectRoles.php:12]: ok

**Comment by rowan04** [comment]
I found that the INSTALL.md information about the sample data is now incorrect: https://github.com/GOCDB/gocdb/blob/f6d4a1b750eecc12c2ca4aa4f38a496154f27b29/INSTALL.md#optional-deploy-sample-data
However it will need to be edited for the remove simple sample data (#415) issue as it also contains information about the simple sample data, so should this file be left for that pull request or should I edit the bit about the normal sample data here?

**Comment by gregcorbett** [comment]
> I found that the INSTALL.md information about the sample data is now incorrect:

Good spot :) 

> should I edit the bit about the normal sample data here?

Yes please, else we run the risk of forgetting about it :)

**Comment by gregcorbett** [comment]
This is looking great 😄, one final thing - cna you remove any `CIC Staff` roles from UsersAndRoles.xml. That is a legacy role that we no longer use.

**Comment by rowan04** [comment]
> This is looking great 😄, one final thing - cna you remove any `CIC Staff` roles from UsersAndRoles.xml. That is a legacy role that we no longer use.

done

**Review by gregcorbett — APPROVED** [review]
LGTM 😄

**Comment by gregcorbett** [comment]
Rebasing

File: lib/Doctrine/deploy/AddNGIs.php (hunk 1 above)
Comment (rowan04) [reply lib/Doctrine/deploy/AddNGIs.php]: done in #428

**Pull request merged**